src/sql/src/plan/statement/dml.rs

```diff
@@ -457,16 +485,88 @@ pub fn plan_subscribe(
                 QueryLifetime::OneShot(scx.pcx()?),
             )?;
             assert!(query.finishing.is_trivial(query.desc.arity()));
-            SubscribeFrom::Query {
-                expr: query.expr,
-                desc: query.desc,
-            }
+            let desc = query.desc.clone();
+            (
+                SubscribeFrom::Query {
+                    expr: query.expr,
+                    desc: query.desc,
+                },
+                desc,
+                query.scope,
+            )
         }
     };
 
     let when = query::plan_as_of(scx, as_of)?;
     let up_to = up_to.map(|up_to| plan_up_to(scx, up_to)).transpose()?;
 
+    let qcx = QueryContext::root(scx, QueryLifetime::OneShot(scx.pcx()?));
+    let ecx = ExprContext {
+        qcx: &qcx,
+        name: "",
+        scope: &scope,
+        relation_type: desc.typ(),
+        allow_aggregates: false,
+        allow_subqueries: true,
+        allow_windows: false,
+    };
+
+    let output = match output {
+        SubscribeOutput::Diffs => plan::SubscribeOutput::Diffs,
+        SubscribeOutput::EnvelopeUpsert { key_columns } => {
+            scx.require_envelope_upsert_in_subscribe()?;
+            let order_by = key_columns
+                .iter()
+                .map(|ident| OrderByExpr {
+                    expr: Expr::Identifier(vec![ident.clone()]),
+                    asc: None,
+                    nulls_last: None,
+                })
+                .collect_vec();
+            let (order_by, map_exprs) = query::plan_order_by_exprs(
+                &ExprContext {
+                    name: "ENVELOPE UPSERT KEY clause",
+                    ..ecx
+                },
+                &order_by[..],
+                &[],
+            )?;
+            if !map_exprs.is_empty() {
+                sql_bail!(
```

Comment:
These two errors should be a new PlanError variant, then the hint/detail fields can be the thing after the semicolon. See https://github.com/MaterializeInc/materialize/blob/main/doc/developer/style.md#error-message-style

Reply:
done